### PR TITLE
Fix route trafficTargets for canary

### PR DESCRIPTION
**Pull request by hougangliu**
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kfserving/336)

### CODE REVIEW

File: pkg/controller/kfservice/resources/knative/route.go [hunk 1]
```diff
@@ -53,6 +56,10 @@ func (r *RouteBuilder) CreateKnativeRoute(kfsvc *v1alpha2.KFService, endpoint co
 		},
 	}
 	if kfsvc.Spec.Canary != nil {
+		endpoint = constants.Predictor
+		if kfsvc.Spec.Canary.Transformer != nil {
```
Comment (rakelkar) [inline pkg/controller/kfservice/resources/knative/route.go]: add a test for this case?

Comment (hougangliu) [reply pkg/controller/kfservice/resources/knative/route.go]: sure, done. Thanks

**Comment by yuzisun** [comment]
@hougangliu @rakelkar I actually left this intentionally and opened issue (https://github.com/kubeflow/kfserving/issues/324) since I am not sure if canarying(add/remove) transformer really makes sense :), I'd like to hear the thoughts from you guys.

**Comment by rakelkar** [comment]
I think it makes sense. When I saw the code it looked like a bug to me too... I think that once you are taking live traffic all changes need to be canaried.

**Comment by hougangliu** [comment]
IMP, we'd better not restrict that `default` must keep same traffic flow as `canary`.
For example, a vendor has v1 model, which only accepts float vector input, but end user only wants to pass a file as argument for predicting, in this case, we should add a transformer before predict. as we all know, transformer will pass its output to predictor server by network, which will take some time, so the vendor maybe develop a new v2 model which can accept a file as input, and after model trained, he wants to make both v1 model and v2 model online while v2 works as canary. in this case, for v1 model, a transformer is in need, but not v2.
So I think keeping it open can benefit.

**Comment by yuzisun** [comment]
cool, totally valid use case but the question is how we can prevent the invalid use cases. If model v1 accepts tensor with only predictor and v2 adds transformer with file input, the requests could fail when the traffic is forwarded to default according to canary split percentage.  I think above use case is also related to the optimization we’d like to do to place transformer onto the same pod then we can leverage shared memory or local network communication.

**Comment by hougangliu** [comment]
@yuzisun  Yes, scheduling transformer pod and predictor pod on same node can improve performance to some extent. But as above use case,  I think v2 model with sufficient CPU/Mem/GPU resource can work better than it, and I believe there is such use case and requirement, and we should support such use case.
Also, we always should come to an agreement that a KFService's default endpoint and canary endpoint have same input and output, and an endpoint is made of combination of explainer, transformer and predictor, and as for how to combine the component, I think it should be decided by the end user (he just need make sure that endpoint and canary endpoint have same input and output)
so for your question how we can prevent the invalid use cases, I believe that we can never handle all the invalid use cases (for example, even we limit that both default and canary endpoint should have same combination, we cannot make sure default and canary endpoint have same input and output)

**Comment by rakelkar** [comment]
> the requests could fail when the traffic is forwarded to default according to canary split percentage

Isn't that the whole point of canary? If you do a breaking change you need to create a new service. Canary helps you verify the change is good.

**Comment by rakelkar** [comment]
/hold

Based on the discussion, we should not make this change (details in #324)

**Comment by hougangliu** [comment]
ok, then I close it based on the decision #324

**Pull request closed without merging**